certbot-nginx/certbot_nginx/challenges.py

```diff
@@ -0,0 +1,330 @@
+"""Classes that perform challenges for Nginx"""
+
+import logging
+import os
+
+import six
+
+from acme import challenges
+
+from certbot import errors
+from certbot.plugins import common
+
+from certbot_nginx import obj
+from certbot_nginx import nginxparser
+
+
+logger = logging.getLogger(__name__)
+
+class NginxChallengePerformer(common.ChallengePerformer):
+    """Additional helper methods for Nginx challenge performers."""
+
+    def perform(self):
+        """Perform all added challenges.
+
+        :returns: challenge respones
+        :rtype: `list` of `acme.challenges.KeyAuthorizationChallengeResponse`
+
+
+        """
+        raise NotImplementedError() # pragma: no cover
+
+    def _make_server_block(self, achall, addrs):
+        """Creates a server block for a challenge.
+
+        :param achall: Annotated HTTP-01 challenge
+        :type achall:
+            :class:`certbot.achallenges.KeyAuthorizationAnnotatedChallenge`
+
+        :param list addrs: addresses of challenged domain
+            :class:`list` of type :class:`~nginx.obj.Addr`
+
+        :returns: server block for the challenge host
+        :rtype: list
+
+        """
+        raise NotImplementedError() # pragma: no cover
+
+    @property
+    def _challenge_conf(self):
+        """Location of the challenge config file"""
+        raise NotImplementedError() # pragma: no cover
+
+    def _listen_addresses(self, default_addr, ipv6_addr, port):
+        """Finds addresses for each challenge block to listen on.
+
+        :param string default_addr: default listen directive argument for ipv4
+        :param string ipv6_addr: default listen directive argument for ipv6
+        :param int port: port to check for ipv6 usage
+
+        :returns: list of lists of :class:`certbot_nginx.obj.Addr` to apply
+        :rtype: list
+
+        """
+        addresses = []
+        ipv6, ipv6only = self.configurator.ipv6_info(port)
+
+        for achall in self.achalls:
+            vhost = self.configurator.choose_vhost(achall.domain, create_if_no_match=True)
+
+            if vhost is not None and vhost.addrs:
+                non_ssl_addrs = (addr for addr in vhost.addrs if not addr.ssl)
```

Comment:
I was looking through this while giving Joona to read the stuff I wrote and I think this line is the problem with TLS-SNI on this branch. Before the refactor, this line was https://github.com/certbot/certbot/blob/9e952081014b9545ce6ddb8b6ecc86a51bf94131/certbot-nginx/certbot_nginx/tls_sni_01.py#L61.

Applying the below diff to this branch fixed the compatibility tests when running them locally:
```
$ git diff
diff --git a/certbot-nginx/certbot_nginx/challenges.py b/certbot-nginx/certbot_nginx/challenges.py
index aff1c57..ee1c17e 100644
--- a/certbot-nginx/certbot_nginx/challenges.py
+++ b/certbot-nginx/certbot_nginx/challenges.py
@@ -68,8 +68,9 @@ class NginxChallengePerformer(common.ChallengePerformer):
             vhost = self.configurator.choose_vhost(achall.domain, create_if_no_match=True)
 
             if vhost is not None and vhost.addrs:
-                non_ssl_addrs = (addr for addr in vhost.addrs if not addr.ssl)
-                addresses.append(list(non_ssl_addrs))
+                #non_ssl_addrs = (addr for addr in vhost.addrs if not addr.ssl)
+                #addresses.append(list(non_ssl_addrs))
+                addresses.append(list(vhost.addrs))
             else:
                 if ipv6:
                     # If IPv6 is active in Nginx configuration
```
Obviously the real fix will be something different, but this should mean we're on the right track!

Reply:
Ah thanks!! I knew there had to be something I'd missed that I'd changed in the refactor...

Reply:
Not that I'm still awake

Reply:
this is a good reminder that more-awake erica should go over all the changes with a fine-tooth comb